effects.cc

```diff
@@ -251,17 +226,35 @@ void Effects_manager::remove_usecode_lightning(
 
 int Effects_manager::get_weather(
 ) {
-	Special_effect *each = effects;
-	while (each) {
-		Special_effect *next = each->next;
-		if (each->is_weather()) {
-			auto *weather = static_cast<Weather_effect *>(each);
-			if (weather->get_num() >= 0)
-				return weather->get_num();
-		}
-		each = next;
-	}
-	return 0;
+	auto found = std::find_if(effects.cbegin(), effects.cend(), [](const auto& ef) {
+			return ef->is_weather() && 
+				(static_cast<Weather_effect *>(ef.get())->get_num() >= 0);
+	});
+	return found != effects.cend()
+		? static_cast<Weather_effect *>((*found).get())->get_num()
+		: 0;
+}
+
+/**
+ * Construct and store game window singleton
+ * TODO check if passing it from other scope
+ * won't be better
+ */
+
+Special_effect::Special_effect(
+) : gwin(Game_window::get_instance())
+{}
+
+/**
+ * Destructor
+ * Takes advantage of the fact that gwin
+ * is stored.
+ */
+Special_effect::~Special_effect(
+)
+{
+	if (in_queue())
+		gwin->get_tqueue()->remove(this);
```

Comment:
I am half-wondering if this code should not be pushed all the way to the Time_sensitive base class; it does not make sense to have to manually remove each and every Time_sensitive by hand when they are deleted. But this can be for a future set of commits, this is just me musing out loud.

Reply:
I'll have a look at it.